components/service-operator/controllers/serviceaccount.go

```diff
@@ -0,0 +1,153 @@
+	Log                  logr.Logger            // Log will be used to report each reconcile
+	KubernetesClient     client.Client          // KubernetesClient is required to talk to api
+}
+
+// SetupWithManager validates and registers this controller with the manager and api
+func (r *ServiceAccountController) SetupWithManager(mgr ctrl.Manager) error {
+	r.KubernetesClient = mgr.GetClient()
+	r.Scheme = mgr.GetScheme()
+	// setup logger
+	r.Log = ctrl.Log.WithName("controllers").WithName((&core.ServiceAccount{}).GetResourceVersion())
+	// register with manager
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&core.ServiceAccount{}).
+		Complete(r)
+}
+
+// Reconcile synchronises state between the resource and a cloudformation stack
+func (r *ServiceAccountController) Reconcile(req ctrl.Request) (res ctrl.Result, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second * 1)
+	defer cancel()
+	// execute reconciliation and log changes
+	op, err := r.reconcileWithContext(ctx, req)
+	if err == context.DeadlineExceeded {
+		// ran out of time, most likely waiting on
+		// a long running provisioning, come back a bit later
+		res.Requeue = true
+		res.RequeueAfter = time.Second * 1
+		err = nil
+	}
+	r.Log.Info("reconciled",
+		"service", req.NamespacedName,
+		"requeue", res.Requeue,
+		"after", res.RequeueAfter,
+		"op", op,
+		"err", err,
+	)
+	return res, err
+}
+
+// reconcileWithContext fetches the resource to reconcile and executes reconcileObjectWithContext and returns if any changes were made
+func (r *ServiceAccountController) reconcileWithContext(ctx context.Context, req ctrl.Request) (controllerutil.OperationResult, error) {
+	bg := context.Background()
+	o := &core.ServiceAccount{}
+	if err := r.KubernetesClient.Get(bg, req.NamespacedName, o); apierrs.IsNotFound(err) {
+		// nothing we can do if the resource has gone missing, so
+		// ignore any not found errors and let the api carry on
+		return controllerutil.OperationResultNone, nil
+	} else if err != nil {
+		// issue communicating with the api
+		// return err and we'll retry later
+		return controllerutil.OperationResultNone, err
+	}
+
+	if _, ok := o.ObjectMeta.Labels[cloudformation.AccessGroupLabel]; !ok {
+		return controllerutil.OperationResultNone, nil
+	}
+
+	// track changes to our object resource and call the main reconcile func
+	var reconcileErr error
+	op, updateErr := controllerutil.CreateOrUpdate(bg, r.KubernetesClient, o, func() error {
+		reconcileErr = r.reconcileObjectWithContext(ctx, req, o)
+		return nil // always try to update
+	})
+	if reconcileErr != nil {
+		return op, reconcileErr
+	} else if updateErr != nil {
+		return op, updateErr
+	}
+	return op, nil
+}
+
+func (r *ServiceAccountController) updatePrincipal(ctx context.Context, o *core.ServiceAccount) (*access.Principal, error) {
+	principal := &access.Principal{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      o.GetName(),
```

Comment:
:thinking: should we use `GenerateName` here instead of `Name` to avoid clashes?

Reply:
Unfortunately we make a sneaky assumption when we generate the assume role policy for the Principal's IAM Role that the ServiceAccount will get the same name as the Principal.
Maybe we should look at the Principal's ownerReference though?

Reply:
We discussed and think the Principal should have a spec field to specify ServiceAccount - if not set, it won't trust anything extra. If set, it will trust the ServiceAccount specified. Then we just set that spec field when creating it in the ServiceAccountController.

Reply:
I've been thinking about this - won't this result in a new Principal every time the ServiceAccount gets reconciled as the name will be randomly chosen each time?